lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -774,8 +824,83 @@ void FIRRTLTypesLowering::getAllBundleLowerings(
       .Default([&](auto) {});
 }
 
-Identifier FIRRTLTypesLowering::getArgAttrName(unsigned newArgNumber) {
+Identifier TypeLoweringVisitor::getArgAttrName(unsigned newArgNumber) {
   SmallString<16> argNameBuffer;
   mlir::impl::getArgAttrName(newArgNumber, argNameBuffer);
-  return Identifier::get(argNameBuffer, &getContext());
+  return Identifier::get(argNameBuffer, context);
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerTypesPass : public LowerFIRRTLTypesBase<LowerTypesPass> {
+  void runOnOperation() override;
+
+private:
+  void runAsync();
+  void runSync();
+  // A list of visitors, one for each thread.  These are cached by the pass.
+  std::vector<TypeLoweringVisitor> visitors;
+};
+} // end anonymous namespace
+
+void LowerTypesPass::runAsync() {
+  // Create a TypeLoweringVisitor for each thread to use.
+  size_t numThreads = llvm::hardware_concurrency().compute_thread_count();
+  visitors.resize(numThreads, &getContext());
+
+  // Used to divide up the work, first come first serve.
+  std::atomic<unsigned> opIt(0);
+
+  // Process the operations in parallel, with one runner for each hardware
+  // thread.  Each thread will get an operation index to process, and walk the
+  // list of operations until it finds that operation. It will repeat this until
+  // there are no more operations in the list.
+  mlir::ParallelDiagnosticHandler diagHandler(&getContext());
+  auto &ops = getOperation().getBody()->getOperations();
+  auto perThread = [&](TypeLoweringVisitor &visitor) {
+    // This is the operation index that this thread should process.
+    auto nextId = opIt++;
+    // This is the current operation index.
+    unsigned curId = 0;
+    for (auto &op : ops) {
+      // Skip forward until we are at operation this thread should process.
+      if (curId++ != nextId) {
+        continue;
+      }
+
+      // Notify the handler the op index and then perform lowering.
+      diagHandler.setOrderIDForThread(curId);
+      visitor.lowerModule(&op);
+      diagHandler.eraseOrderIDForThread();
+
+      // Get the next op to process.
+      nextId = opIt++;
+    }
+  };
+  llvm::parallelForEach(visitors.begin(), visitors.end(), perThread);
 }
+
+void LowerTypesPass::runSync() {
+  auto circuit = getOperation();
+  TypeLoweringVisitor visitor(&getContext());
+  for (auto &op : circuit.getBody()->getOperations()) {
+    visitor.lowerModule(&op);
+  }
+}
+
+// This is the main entrypoint for the lowering pass.
+void LowerTypesPass::runOnOperation() {
+  if (getContext().isMultithreadingEnabled()) {
```

Comment:
Please check, but I'd expect `parallelForEach` to turn serial when multithreading isn't enabled.

Reply:
It will turn serial if LLVM is compiled without multithreading support, but that is different than respecting the `-mlir-disable-threading` runtime flag.

Reply:
got it, thanks!